suggested fix for data-dependent error

avikchaudhuri opened the pull request:
Suggests fixes for data-dependent errors in non-strict export.

Any data-dependent error has an unresolved condition on unbacked symints. A mechanizable strategy for fixing such errors, which this PR enables, is to "bash" them using `torch._check()`s. For each error we suggest using `torch._check()` on the condition or its negation. The user selects and copy-pastes the suggested fix and continues.

For example, here's an existing data-dependent error message with the suffix following `<snip>...</snip>` added by this PR:
```
Could not guard on data-dependent expression Eq(u2, u1) (unhinted: Eq(u2, u1)).  (Size-like symbols: u1)

<snip>...</snip>

User code:
  File "test/export/test_export.py", line 1944, in forward
    return r.view(items[0], items[2])

Suggested fixes (please choose one of the following):
  1. torch._check(items[2] == r.shape[1])
  2. torch._check(items[2] != r.shape[1])"
```

Tests in this PR illustrate this workflow, by taking common examples of data-dependent errors and bashing them until success, purely based on suggested fixes. In particular, we test this workflow on the "puzzlers" in https://www.internalfb.com/intern/anp/view/?id=5330476 (thanks @ezyang).

In terms of implementation, we focus on non-strict mode, where we can intercept torch function calls to install a handler that walks up the stack from the error, finding the closest non-torch frame and inspecting its locals for symints appearing in the error. The suggested fixes then access these symints through the local variables so that they can be (a) easily understood by the user (b) directly added to the code.

Implementing this idea in strict mode is follow-up work—we have already investigated what it would take, and decided to separate it out of this PR for reasons described next. 

It's not too hard to map symints to locals in Dynamo (although it needs to happen elsewhere, i.e., intercepting torch function calls won't work). However, unfortunately this doesn't seem to be enough; the graph modules created by Dynamo when going through AOTAutograd can raise further data-dependent errors in some cases, and thus we need yet another mechanism to map symints to locals for graph modules, via captured source-level metadata and FX node walking. This latter component will require some care to build properly, or we might conclude it is altogether unnecessary and fix Dynamo instead.

Differential Revision: D56867432


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

avikchaudhuri commented:
failing tests because https://www.internalfb.com/diff/D59561509 is out of sync

ezyang commented:
Although the regexes are good for unit tests, it would be nice to see the actual, literal messages that are emitted in the PR desc

avikchaudhuri commented:
> Although the regexes are good for unit tests, it would be nice to see the actual, literal messages that are emitted in the PR desc

Updated the PR description.

ezyang commented:
It's hard to say without using it in practice, but I kind of wonder if we should be directly suggesting the torch._check, or if we should just be offering the symbolic sizes of bindings so that the user can pick one that's semantically appropriate. I worry that we might pick some technically correct but difficult to understand spelling of a symbolic shape when the user can select a better one. This is something I'm happy to test on the road.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
+        return sympy.floor(x)
+
+    # find the unresolved condition on unbacked symints in the error message
+    match = re.search(r"data-dependent expression (.+) \(unhinted", e.args[0])
```
pretty meh about this regexing

ezyang replied:
We control the exception type. So we can just retain the sympy expression directly instead of parsing it out of the message

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
+    frame = inspect.currentframe()
+    while frame is not None:
+        # walk the stack up from the data-dependent error until a non-torch frame is found
+        if not frame.f_code.co_filename.startswith(os.path.dirname(inspect.getfile(torch))):
```
This is one of the things I worry about, the first non-torch frame may not actually be the correct place to put the assert, if another library like torchrec is involved

ezyang replied:
Also, this test will match torchrec (because you'll end up with something like "/blah/blah/torch" from the path here, matching "/blah/blah/torchrec/...")

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
+                # figure out how to access any symbol inside `val` through `var`
+                for path, leaf in pytree.tree_leaves_with_path(val):
+                    name = var + pytree.keystr(path)
+                    if isinstance(leaf, torch.SymInt):
```
Introspecting into list-y/tuple-y things is probably the most obvious other thing I would do here

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
+    """
+    Util printer that replaces sympy symbols with their source-level names
+    and renders sympy relational operators (e.g., Eq, Ne, Ge, Le) inline
+    (i.e., as ==, !=, >, <).
```
It's not obvious to me that this handles all of our sympy functions like PythonMod correctly

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
+            a, b = x.args
+            if isinstance(a, sympy.Rational):
+                return FloorDiv(a.p * b, a.q)
+        return sympy.floor(x)
```
And if you retain the orig expression you don't need this fixer

avikchaudhuri replied:
Sorry, missed this message before landing. Follow up here: #131932

ezyang reviewed (approved):
Seems good enough for testing

avikchaudhuri committed:
suggested fix for data-dependent error (#125378)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/125378

Test Plan: added test

Reviewed By: angelayi

Differential Revision: D56867432
